### ISSUE TITLE
Element not properly highlighted

### ISSUE THREAD

**Issue by thomasledoux1** [benmarch/angular-ui-tour#163]
* Browser: Chrome Version: 62.0.3202.94
* AngularJS: ~1.4
* Angular Bootstrap: ~3
* Angular UI Tour: ^0.9.2

**I have installed this library via:** (NPM, Bower, or downloaded package): Bower

**I have observed the following behavior:**
When I want to highlight an element and have the rest of the page have a backdrop, the highlighted area is around 15 pixels under the element which should be highlighted. 

**This is how I expected it to behave:**
I expected the highlighted element which has the tour-step on it to have a white background (to be highlighted) and the rest of the body of my website to have a backdrop.

**Here is my tour config, and all related step configs:**

```js
var tourConfig = {
     name: 'general',
     debug: true,
     useHotkeys: true,
     templateUrl: 'modules/players/views/templates/tour-step-template.html',
     orphan: true,
     useUiRouter: true,
     backdrop: true
};
```

Tour step: 
```html
<button class="btn btn-primary button-wrapper" data-ng-show="!vm.isParent" ui-sref="players.connect" tour-step="kindToevoegen" tour-step-placement="right" tour-step-orphan="false" tour-step-backdrop="true" tour-step-append-to-body="true" tour-step-next-path="players.connect" tour-step-belongs-to="general" tour-step-next-step="testNewPage" tour-step-order="1" tour-step-content="<p style='text-align: center;'>Klik op 'Kind toevoegen' om een nieuw kind toe te voegen aan de Karaton Academy.</p>">Kind toevoegen</button>
```

**Additional notes/code:**
When I delete the css classes from the button, the effect stays the same. 
See screenshot for the result.

![screen shot 2017-11-23 at 09 43 17](https://user-images.githubusercontent.com/10978632/33164376-c3a14258-d032-11e7-83e5-12ad105c2935.png)

**Comment by benmarch** [comment]
Hey @thomasledoux1, if you inspect the step target (your element) in dev tools, does it highlight the original button or the region that the backdrop is highlighting? Can you take a screenshot of the highlighted element?

**Comment by thomasledoux1** [comment]
Hey @benmarch, the whole body gets highlighted when I inspect the white area in Chrome. I can't inspect the button itself (which has the tour step added to it in HTML) because the backdrop covers it.. 
![screen shot 2017-11-28 at 11 24 39 2](https://user-images.githubusercontent.com/10978632/33314701-bf766af0-d42e-11e7-942a-44706c4ec6f3.png)
On my <html> element I don't have any padding/margin set. On <body> I have padding: 0 and  margin: 0 set.

**Comment by benmarch** [comment]
Hey @thomasledoux1, one thing I can see is that the body doesn't reach the end of the content. That is usually a sign of absolutely positioned or floating elements without a clearfix, and could result in the misalignment. I can't see the computed styles for body in this shot but it doesn't look like there is padding or margin which is good. Can you try to drill down to the target element in the inspector and take another screenshot, or see if you can get the body to fully cover the content?

**Comment by thomasledoux1** [comment]
Hey @benmarch, I tried to fix the body to cover the whole content, but was not able to do so so far. I can see that from the angular-material.css it's inheriting 'height: 100%' property. When I overwrite this in my core.css with 'height: auto' I get the same result. I tried adding clearfix to the floating elements and body too, but without effect. I have added the computed styles and the generated HTML of the targeted element. 
![screen shot 2017-11-29 at 09 36 12](https://user-images.githubusercontent.com/10978632/33365541-d1a0cfbc-d4e8-11e7-9889-28a5b6b284dc.png)
![screen shot 2017-11-29 at 09 36 01](https://user-images.githubusercontent.com/10978632/33365542-d1bfdb00-d4e8-11e7-8da9-350c7f16fce5.png)
![screen shot 2017-11-29 at 09 35 53](https://user-images.githubusercontent.com/10978632/33365543-d1d428da-d4e8-11e7-9515-18518f79fa4d.png)

**Comment by benmarch** [comment]
Can you take a screenshot of the page with the target button highlighted and backdrop showing? I'm looking to see where the highlight is relative to the backdrop.

**Comment by thomasledoux1** [comment]
@benmarch here you go! 
![screen shot 2017-11-30 at 09 48 05](https://user-images.githubusercontent.com/10978632/33421806-9c636c8a-d5b3-11e7-8fe2-a358eea9acce.png)

**Comment by benmarch** [comment]
Thanks, @thomasledoux1, this confirms that it is definitely an issue with my positioning logic. Is there any way I can debug this using your source code?

Out of curiosity, does the issue resolve itself if you slightly resize the window (even just by a pixel)? And how did you get the number buttons ("1" on the left, "10" on the right) to highlight in addition to the button?

**Comment by thomasledoux1** [comment]
Hey @benmarch, resizing the window doesn't change this situation. I tried all different sizes but no effect. I think the buttons are highlighted because they had the same grey as the backdrop, so they change to white to be visible? Is that possible? 

This is all ran local, I can't put it on our staging environment at the moment, so I can't give you access..

**Comment by benmarch** [comment]
Sorry for the delay... if you inspect the backdrop, is it a child of the body? I find it strange that you have elements in front of it, although I suppose that could be a z-index property. It also looks like the alignment is off by approximately the height of your header, so I wonder if that is pushing it down or affecting the calculation somehow.

**Comment by thomasledoux1** [comment]
Hey @benmarch no problem! The backdrop is a part of the body, as you can see in the attached screenshot. I also found it strange that some elements pop up in front of the backdrop, but I have no idea what is causing that.. The elements have the 'active' class on them, might that cause a problem? (see screenshot 2). It could be caused by the header, but how can I influence the calculation then?
![screen shot 2017-12-05 at 16 42 35](https://user-images.githubusercontent.com/10978632/33615860-807845d2-d9db-11e7-9fd1-91b2fb0fd154.png)
![screen shot 2017-12-05 at 16 40 32](https://user-images.githubusercontent.com/10978632/33615861-8090b3ec-d9db-11e7-91bd-e18a6f719699.png)

**Comment by benmarch** [comment]
Thanks for the screenshots @thomasledoux1. I'm not sure if I will be able to debug this without looking at your code, but I can explain what I would do if I could:

- The first thing I would check is to see if your `<html>` or `<body>` elements have any padding or margin, or if the `box-sizing` is anything but `border-box`.
- It's very strange that you have some elements that are poking through the backdrop. It makes me think that there is some CSS that defines them with a high `z-index` but no `position: relative`, or possibly some malformed HTML. If you can get the backdrop to display with no elements poking through, it might inadvertently fix the issue.
- The last thing is that the `<body>` does not enclose all of its content. This is a symptom of floating of elements whose parents have not been clearfixed. This is a common issue that can cause the position to be calculated incorrectly. Unfortunately, this can be a difficult one to fix because often the elements are deeply nested in the DOM tree.

Try playing around with the `<body>` and `<html>` box styles, and the backdrop `z-index`. I think it will come down to something in there. Sorry I can't be of more help.

**Comment by thomasledoux1** [comment]
@benmarch The code with the UI Tour implemented is now on our staging environment, I can give you access with an account. Is there any way I can send you the details for it in private?

**Comment by benmarch** [comment]
Hey @thomasledoux1, ping me at b89spam@gmail.com and I'll reply from my personal email.

**Comment by thomasledoux1** [comment]
Hey @benmarch, I sent you an e-mail!

**Comment by benmarch** [comment]
Hey @thomasledoux1, sorry for the delay, but I think I can solve your issue. There are a couple minimal changes that you can make, plus I have a few recommendations for you that will help avoid these kinds of issues in the future. There are screenshots for these below:
- Add the `clearfix` class to your header component (will fix the positioning)
- Override the angular-material styles for `html, body` and set `height: auto; min-height: 100%` for both html and body (will fix the coverage) 
- Recommendations:
  - Your nav is a fixed element inside your header component which is static. I recommend pulling the fixed element up to be a child of the body element. This isn't necessary, but it will be clearer when debugging styles
  - There are multiple elements that have only floated children but are not clearfixed. These elements show up as 0 height, but have non-zero-height content. Add the clearfix class to these.
  - There is at least one element with both a `row` class and a `col-xx-x` class. It should really be one or the other (I think its the menu)

#  Screenshots:
## Header change:
![screen shot 2017-12-21 at 2 45 35 pm](https://user-images.githubusercontent.com/1646126/34273725-8d772df4-e663-11e7-98f1-7066023d1a59.png)

## Body Changes

### Before
![screen shot 2017-12-21 at 3 19 56 pm](https://user-images.githubusercontent.com/1646126/34273738-9f83b2e2-e663-11e7-9662-896e86cb756c.png)

### After
![screen shot 2017-12-21 at 3 20 48 pm](https://user-images.githubusercontent.com/1646126/34273749-a6af48d8-e663-11e7-98c1-95440d726fe2.png)

**Comment by dshk** [comment]
Hi @benmarch !
I have the same situation but the issue resolves if I resize the window (even just by a pixel)? 
What should I do to fix it?

**Comment by benmarch** [comment]
Hey @dshk and @thomasledoux1, I have released version 0.9.4 that enables a new option for your tour/tour step: `disableBackdropOptimizations` (see README) that should help with your issue. You might also have to call `tour.reposition()` after the step loads depending on your exact issues. I will work on fixing that as well for the next release.
